packages/host/tests/helpers/adapter.ts

```diff
@@ -0,0 +1,299 @@
+        throw new Error(`tried to use file as directory`);
+      }
+      this.#lastModified.set(this.#paths.fileURL(path).href, now);
+      dir.contents[last] = { kind: 'file', content };
+    }
+  }
+
+  createJWT(claims: TokenClaims, expiration: string, secret: string) {
+    return createJWT(claims, expiration, secret);
+  }
+
+  verifyJWT(
+    token: string,
+    secret: string,
+  ): TokenClaims & { iat: number; exp: number } {
+    let [_header, payload, signature] = token.split('.');
+    if (signature === secret) {
+      let claims = JSON.parse(atob(payload)) as {
+        iat: number;
+        exp: number;
+      } & TokenClaims;
+      let expiration = claims.exp;
+      if (expiration > Date.now() / 1000) {
+        throw new TokenExpiredError(`JWT token expired at ${expiration}`);
+      }
+      return claims;
+    }
+    throw new JsonWebTokenError(`unable to verify JWT: ${token}`);
+  }
+
+  get lastModified() {
+    return this.#lastModified;
+  }
+
+  // this is to aid debugging since privates are actually not visible in the debugger
+  get files() {
+    return this.#files;
+  }
+
+  async *readdir(
+    path: string,
+  ): AsyncGenerator<{ name: string; path: string; kind: Kind }, void> {
+    let dir =
+      path === '' ? this.#files : this.#traverse(path.split('/'), 'directory');
+    for (let [name, content] of Object.entries((dir as Dir).contents)) {
+      yield {
+        name,
+        path: path === '' ? name : `${path}/${name}`,
+        kind: content.kind,
+      };
+    }
+  }
+
+  async exists(path: string): Promise<boolean> {
+    let maybeFilename = path.split('/').pop()!;
+    try {
+      // a quirk of our test file system's traverse is that it creates
+      // directories as it goes--so do our best to determine if we are checking for
+      // a file that exists (because of this behavior directories always exist)
+      await this.#traverse(
+        path.split('/'),
+        maybeFilename.includes('.') ? 'file' : 'directory',
+      );
+      return true;
+    } catch (err: any) {
+      if (err.name === 'NotFoundError') {
+        return false;
+      }
+      if (err.name === 'TypeMismatchError') {
+        try {
+          await this.#traverse(path.split('/'), 'file');
+          return true;
+        } catch (err: any) {
+          if (err.name === 'NotFoundError') {
+            return false;
+          }
+          throw err;
+        }
+      }
+      throw err;
+    }
+  }
+
+  async openFile(
+    path: LocalPath,
+    loader: Loader,
+  ): Promise<FileRef | undefined> {
+    let content;
+    try {
+      content = this.#traverse(path.split('/'), 'file');
+    } catch (err: any) {
+      if (['TypeMismatchError', 'NotFoundError'].includes(err.name)) {
+        return undefined;
+      }
+      throw err;
+    }
+    if (content.kind === 'directory') {
+      return undefined;
+    }
+
+    let value = content.content;
+
+    let fileRefContent = '';
+
+    if (path.endsWith('.json')) {
```

Comment:
We are now serializing lazily - happens only when the file that lives in the test adapter is actually requested.